### PR TITLE
Add Basic Auth support to Sawtooth CLI

### PR DESCRIPTION
**Pull request by delventhalz**
# To Test This PR

This PR adds support for Basic Auth proxies. In order to test it, you'll need to build one. These are the instructions for setting up a simple authed Apache proxy in vagrant.

## Install Apache

```
% sudo apt-get update
% sudo apt-get install -y apache2
% sudo a2enmod headers
% sudo a2enmod proxy_http
```

## Set up a password file
```
% vi /tmp/.password
```

Edit the file to look like this:
```
sawtooth:$apr1$VHIPQ4Wi$k7PbIPqcfVp1NVKWhkIbS0

```

This saves the password `"sawtooth"` for the user `sawtooth`. You can generate new files like this at the site: http://www.htaccesstools.com/htpasswd-generator/. However, the config file below only allows the user `sawtooth`, so that would need to be modified as well.

## Set up the site configuration

```
% sudo vi /etc/apache2/sites-enabled/000-default.conf
```

Edit the file to look like this:
```
<VirtualHost *:80>
        ServerAdmin sawtooth@sawtooth
        DocumentRoot /var/www/html

        ErrorLog ${APACHE_LOG_DIR}/error.log
        CustomLog ${APACHE_LOG_DIR}/access.log combined

        <Location />
                Options Indexes FollowSymLinks
                AllowOverride None
                AuthType Basic
                AuthName "Enter password"
                AuthUserFile "/tmp/.password"
                Require user sawtooth
                Require all denied
        </Location>
</VirtualHost>

ProxyPass /sawtooth http://localhost:8080
ProxyPassReverse /sawtooth http://localhost:8080
RequestHeader set X-Forwarded-Path "/sawtooth"
```

## Start and/or restart Apache

```
% sudo apachectl start
```

```
% sudo apachectl restart
```

## Make sure everything is set up properly

Start up a validator and the REST API normally. When you query the REST API, you should get back some block data as JSON:
```
% curl localhost:8080/blocks
```

You should also be able to get a `401` by querying the proxy without authorization:
```
% curl localhost/sawtooth/blocks
```

And finally, you should be able to get the same JSON by querying the proxy with authorization:
```
% curl localhost/sawtooth/blocks -u sawtooth:sawtooth
```

## Test the sawtooth CLI

With everything working above, you should be able to:
```
sawtooth blocks list --url http://localhost/sawtooth -u sawtooth:sawtooth
```

## What to do if you have permissioning issues around port 80

Sometimes vagrant seems to get angry when you try to run the proxy on port `80`. You can switch it to port `8000` by changing two files.
```
% sudo vi /etc/apache2/ports.conf
```

Switch `Listen 80` -> `Listen 8000`.

```
% sudo vi /etc/apache2/sites-enabled/000-default.conf
```

Switch `<VirtualHost *:80>` -> `<VirtualHost *:8000>`

Then:
```
% sudo apachectl restart
```

Afterwards the above example should work fine, but the proxy url will be `localhost:8000/sawtooth` instead of `localhost/sawtooth`.

### CODE REVIEW

File: cli/sawtooth_cli/batch.py [hunk 1]
```diff
@@ -17,6 +17,9 @@
 from sawtooth_cli import format_utils as fmt
 from sawtooth_cli.rest_client import RestClient
 from sawtooth_cli.exceptions import CliException
+from sawtooth_cli.parent_parsers import rest_parent
```
Comment (peterschwarz) [inline cli/sawtooth_cli/batch.py]: These names are a little too generic, which made it difficult to figure out what they were for when reading the code.  

Suggest: `general_rest_args_parent`, `list_args_parent`, etc.

**Comment by peterschwarz** [comment]
tested - works with a user in the `.password` file named "sawtooth", based on the instructions given.

File: cli/sawtooth_cli/batch.py (hunk 1 above)
Comment (delventhalz) [inline cli/sawtooth_cli/batch.py]: `rest_parent_parser`?

File: cli/sawtooth_cli/parent_parsers.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        '-u', '--user',
+        type=str,
+        metavar='USERNAME[:PASSWORD]',
+        help='user login info to authorize request')
+
+    return base_parser
+
+
+def base_list_parser():
+    """Creates a parser with arguments specfic to formatting lists
```
Comment (ojalatodd) [inline cli/sawtooth_cli/parent_parsers.py]: Typo in comments: specfic -> specific

File: cli/sawtooth_cli/parent_parsers.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        action='store',
+        default='default',
+        choices=['csv', 'json', 'yaml', 'default'],
+        help='the format to use when printing the output')
+
+    return base_parser
+
+
+def base_show_parser():
+    """Creates a parser with arguments specfic to formatting a single resource.
```
Comment (ojalatodd) [inline cli/sawtooth_cli/parent_parsers.py]: typo: specfic

**Pull request merged**